### PR TITLE
Add executing under test behavior

### PR DESCRIPTION
**Pull request by derekparker**
Details of "why" are in the last commit, copied here for convenience:

This patch alters the code to fallback to non-FIPS compliant routines
when we detect we are running under a test. This is to allow for us to
use this backend with the Go standard library and also pass all the
tests without invasive modifications to the Go stdlib test functions to
ignore or skip unsupported configurations.

### CODE REVIEW

**Review by simo5 — APPROVED** [review]
LGTM

File: openssl/goopenssl.h [hunk 1]
```diff
@@ -726,6 +734,7 @@ DEFINEFUNCINTERNAL(int, RSA_pkey_ctx_ctrl,
 		   (EVP_PKEY_CTX *ctx, int optype, int cmd, int p1, void *p2),
 		   (ctx, optype, cmd, p1, p2))
 
+    #include <openssl/rsa.h>
```
Comment (ueno) [inline openssl/goopenssl.h:737]: This is already include at https://github.com/golang-fips/openssl-fips/pull/10/files#diff-fd756b4114a2dc406db5c6fe6e0aa83892568d7536d6944decc24be096657067R479

File: openssl/openssl_evp.c [hunk 2]
```diff
@@ -38,6 +38,38 @@ int _goboringcrypto_EVP_sign(EVP_MD *md, EVP_PKEY_CTX *ctx, const uint8_t *msg,
...
+  if (!ctx && !(ctx = _goboringcrypto_EVP_PKEY_CTX_new(pk, NULL)))
+    goto err;
+
+  if (1 != _goboringcrypto_EVP_PKEY_sign_init(ctx))
+    goto err;
+
+  if (_goboringcrypto_EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PADDING) <= 0)
+    goto err;
+
+  if (1 != _goboringcrypto_EVP_PKEY_sign(ctx, NULL, slen, msg, msgLen))
```
[the remark this answers is not in the document]
Comment (derekparker) [reply openssl/openssl_evp.c]: Done.

File: openssl/openssl.go [hunk 3]
```diff
@@ -147,12 +157,38 @@ func NewOpenSSLError(msg string) error {
...
+// when FIPS mode. It panics only when
+// the system is in FIPS mode.
+func Unreachable() {
+	if Enabled() {
+		panic("openssl: invalid code execution")
+	}
+}
+
+// UnreachableExceptTests marks code that should be unreachable
+// when FIPS mode is active. It panics.
```
[the remark this answers is not in the document]
Comment (derekparker) [reply openssl/openssl.go]: Done.

File: openssl/goopenssl.h (hunk 1 above)
Comment (derekparker) [reply openssl/goopenssl.h]: Done.

File: openssl/aes_test.go [hunk 4]
```diff
@@ -6,10 +6,14 @@ package openssl
 import (
 	"bytes"
 	"crypto/cipher"
+	"os"
 	"testing"
 )
 
 func TestNewGCMNonce(t *testing.T) {
+	os.Setenv(GoStrictFipsEnv, "1")
+	defer os.Setenv(GoStrictFipsEnv, "0")
```
[the remark this answers is not in the document]
Comment (derekparker) [reply openssl/aes_test.go]: Done.

**Pull request merged**